### PR TITLE
Uninit mirror

### PR DESCRIPTION
**Pull request by ibaned**
Continuation of #1766, contains @dholladay00 's contributions plus some fixes to the unit test.

### CODE REVIEW

**Comment by ibaned** [comment]
Addresses #1765

File: core/unit_test/TestViewAPI.hpp [hunk 1]
```diff
@@ -827,6 +827,17 @@ struct TestViewMirror
...
+  template< class MemoryTraits >
+  void static test_mirror_no_initialize() {
+    Kokkos::View< double*, Layout, Kokkos::HostSpace > a_org( "A", 10 );
+    Kokkos::View< double*, Layout, Kokkos::HostSpace, MemoryTraits > a_h;
+    auto a_d = Kokkos::create_mirror_view( DeviceType(), a_h, Kokkos::WithoutInitializing );
+    
+    int equal_ptr_h_d = (a_h.data() == a_d.data()) ? 1 : 0;
+    constexpr int is_same_memspace = std::is_same< Kokkos::HostSpace, typename DeviceType::memory_space >::value ? 1 : 0;
+    
+    ASSERT_EQ( equal_ptr_h_d, is_same_memspace);
```
Comment (mhoemmen) [inline core/unit_test/TestViewAPI.hpp]: Doesn't `ASSERT_EQ` work with `bool`?

Comment (ibaned) [reply core/unit_test/TestViewAPI.hpp]: Yea, I did bool at first but did this for consistency with surrounding code with uses ints. Kindof pointless, but if we switch this to bool we might as well clean that up in the whole file(s).

**Comment by ndellingwood** [comment]
@ibaned did you run a spot-check on this? I can get one going if you like. 
Spoke to another interested in this feature @DrBooom

**Comment by ibaned** [comment]
@ndellingwood if you don't mind starting one, that would be great

**Comment by ndellingwood** [comment]
@ibaned spot-check running on kokkos-dev

**Comment by ndellingwood** [comment]
@ibaned tests failed with cuda/8.0.44 on kokkos-dev:

```
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-release build_time=187 run_time=393
gcc-5.3.0-OpenMP-release build_time=176 run_time=108
gcc-7.3.0-Serial-release build_time=150 run_time=181
intel-17.0.1-OpenMP-release build_time=432 run_time=163
#######################################################
FAILED TESTS
#######################################################
cuda-8.0.44-Cuda_OpenMP-release (test failed)
```

Failed tests:
```
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] cuda.view_api_b
[  FAILED  ] cuda_uvm.view_api_b
[  FAILED  ] cuda_hostpinned.view_api_b
```

view_api_b output:
```
[ RUN      ] cuda.view_api_b
/home/ndellin/kokkos/core/unit_test/TestViewAPI.hpp:839: Failure
Value of: is_same_memspace
  Actual: 0
Expected: equal_ptr_h_d
Which is: 1
// repeated
```

**Comment by crtrott** [comment]
Jeff will do this.

**Comment by jeffmiles63** [comment]
I think the problem is actually the test case.  
[Kokkos::View< double*, Layout, Kokkos::HostSpace, MemoryTraits > a_h] without assigning it to something or initializing it (via constructor) is a view of size "0".  The mirror [auto a_d = Kokkos::create_mirror_view( DeviceType(), a_h, Kokkos::WithoutInitializing );] will also be a view of size "0".  It appears that in all cases (regardless of memory space), data() returns "0" for a view of size "0".  Thus, the check [int equal_ptr_h_d = (a_h.data() == a_d.data()) ? 1 : 0;] thinks that the data() pointers are the same because they both return "0".
I changed the test case to assign a_h to a_org before creating the mirror; this gives it a size of 80.  I also added a functor and parallel_for to test that the mirror view is accessible on the device.

**Comment by jeffmiles63** [comment]
here is the apollo spot check from the latest code change:

ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'git'
Running on machine: apollo
WARNING!! THE FOLLOWING CHANGES ARE UNCOMMITTED!! :
 M core/unit_test/standalone/Makefile
?? scripts/testing_scripts/test_all.txt

Repository Status:  a6218af9736dc4e2a05d53b3684d8750cd9f931a Merge remote-tracking branch 'origin/develop' into uninit_mirror


Going to test compilers:  gcc/4.8.4 gcc/5.3.0 intel/16.0.1 clang/3.9.0 clang/6.0 cuda/9.1
Testing compiler gcc/4.8.4
  Starting job gcc-4.8.4-OpenMP-release
  PASSED gcc-4.8.4-OpenMP-release
Testing compiler gcc/5.3.0
  Starting job gcc-4.8.4-Pthread-release
  PASSED gcc-4.8.4-Pthread-release
Testing compiler intel/16.0.1
  Starting job gcc-5.3.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
Testing compiler clang/3.9.0
  Starting job intel-16.0.1-OpenMP-release
  PASSED intel-16.0.1-OpenMP-release
Testing compiler clang/6.0
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
  Starting job clang-6.0-Cuda_Pthread-release
  PASSED clang-6.0-Cuda_Pthread-release
Testing compiler cuda/9.1
  Starting job clang-6.0-OpenMP-release
  PASSED clang-6.0-OpenMP-release
  Starting job cuda-9.1-Cuda_OpenMP-release
  PASSED cuda-9.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=128 run_time=419
clang-6.0-Cuda_Pthread-release build_time=222 run_time=285
clang-6.0-OpenMP-release build_time=113 run_time=81
cuda-9.1-Cuda_OpenMP-release build_time=306 run_time=172
gcc-4.8.4-OpenMP-release build_time=105 run_time=104
gcc-4.8.4-Pthread-release build_time=95 run_time=310
gcc-5.3.0-Serial-release build_time=112 run_time=163
intel-16.0.1-OpenMP-release build_time=328 run_time=140

**Comment by jeffmiles63** [comment]
Last thing...I did have to merge my local branch with develop because of the Volta issues fixed in #1973

**Comment by ndellingwood** [comment]
@crtrott this is ready for merge now that the promotion is complete.

**Comment by ndellingwood** [comment]
@DrBooom this is merged into kokkos' develop branch now if you're interested in its use. 
To use the new specialization you need to pass a memory space as the first argument, so something like this:
// Assuming you created some view named "v" with default memory space
// Now we create a host mirror, but with the host view uninitialized
`auto hv = Kokkos::create_mirror_view( Kokkos::HostSpace(), v);`

**Pull request merged**